File: packages/shared/types/response.ts
export type HTTPError = {
  code: number;
  message: string;
};

export type ApiResponse<T> = Promise<
  { data: T; error: undefined } | { data: undefined; error: HTTPError }
>;

// Direct DB operations only
export type AsyncResponseType<T> = Promise<
  { data: T; error: undefined } | { data: undefined; error: Error }
>;

export type SyncResponseType<T> =
  | { data: T; error: undefined }
  | { data: undefined; error: Error };

File: apps/backend-python/src/shared/types/response.py
from typing import Any, Dict, Generic, Optional, TypeVar, Union
from dataclasses import dataclass

from pydantic import BaseModel

T = TypeVar("T")


@dataclass
class HTTPError:
    code: int
    message: str


class ApiResponse(BaseModel, Generic[T]):
    data: Optional[T] = None
    error: Optional[HTTPError] = None


HTTPErrorResponse = Union[tuple[T, None], tuple[None, HTTPError]]

ErrorResponse = Union[tuple[T, None], tuple[None, Exception]]

AuthMiddlewareResponse = HTTPErrorResponse[Dict[str, Any]]
